### PR TITLE
[spmd] switch device mesh collective to use mesh dim groups

### PR DESCRIPTION
**Pull request by wanchaol**
This PR:
* switches device mesh to use dim groups on all collectives
by specifying the `mesh_dim` argument in the collective wrapper calls.
* It switches `scatter` to use c10d collective directly, as the `nn.functional.scatter` have some bugs, we will need to follow up separately. 
* It also deal with the logic that src rank based collective need to
pass in global rank instead of local rank
* It also add some simple example tests on 1d mesh, testing 2d/multi-dim mesh
will come as separate PRs

### CODE REVIEW

File: spmd/tensor/device_mesh.py [hunk 1]
```diff
@@ -211,44 +217,78 @@ def get_rank(self) -> int:
         return get_rank()
 
     def scatter(
-        self, tensors: List[torch.Tensor], src: int = 0
+        self, scatter_list: List[torch.Tensor], mesh_dim: int = 0
     ) -> torch.Tensor:
-        return scatter(tensors, src=src)
+        to_scatter = [tensor.contiguous() for tensor in scatter_list]
```
Comment (fduwjj) [inline spmd/tensor/device_mesh.py:222]: So does this solve the scatter errors observed in rowwise sharding of Linear module?

**Review by fduwjj — APPROVED** [review]
Nice!!! Thanks for fixing that. This is indeed needed right now.

File: spmd/tensor/device_mesh.py (hunk 1 above)
Comment (wanchaol) [reply spmd/tensor/device_mesh.py:222]: Yep I manually checked it, this should solve it, the `nn.functional.scatter` have two bugs:
1. no contiguous call
2. checking local rank instead of global rank
we gonna need to fix it in the core

File: spmd/tensor/device_mesh.py [hunk 2]
```diff
@@ -211,44 +217,78 @@ def get_rank(self) -> int:
         return get_rank()
 
     def scatter(
```
Comment (aazzolini) [inline spmd/tensor/device_mesh.py:219]: Maybe write a docstring to clarify semantic.

**Review by aazzolini — APPROVED** [review]
Looks great!

File: spmd/tensor/device_mesh.py (hunk 2 above)
Comment (wanchaol) [reply spmd/tensor/device_mesh.py:219]: added a docstr for `scatter`, will add more docstr for other collectives in following PRs.
